exercises/grep/.meta/src/reference/java/GrepTool.java

```diff
@@ -0,0 +1,62 @@
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.Collections;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+class GrepTool {
+
+    String grep(String pattern, List<String> flags, List<String> files) {
+        boolean shouldAddFilename = files.size() > 1;
+        return files
+            .stream()
+            .map(file -> grepFile(pattern, flags, file, shouldAddFilename))
+            .filter(line -> !line.isEmpty())
+            .collect(Collectors.joining("\n"));
+    }
+
+    private String grepFile(String pattern, List<String> flags, String filename, boolean shouldAddFilename) {
+        List<String> lines = readFile(filename);
+        String matchingLines = IntStream.range(0, lines.size())
+            .filter(lineIndex -> lineMatchesPattern(lines.get(lineIndex), pattern, flags))
+            .mapToObj(lineIndex -> applyFlagsToLine(lines.get(lineIndex), flags, lineIndex + 1))
+            .map(line -> shouldAddFilename ? filename + ":" + line : line)
+            .collect(Collectors.joining("\n"));
+
+        boolean shouldPrintFilenames = flags.contains("-l");
+        return shouldPrintFilenames && !matchingLines.isEmpty() ? filename : matchingLines;
+    }
+
+    private boolean lineMatchesPattern(String line, String pattern, List<String> flags) {
+        boolean isCaseInsensitive = flags.contains("-i");
+        String lineToCompare = isCaseInsensitive ? line.toLowerCase() : line;
+        String patternToCompare = isCaseInsensitive ? pattern.toLowerCase() : pattern;
+
+        boolean shouldMatchEntireLine = flags.contains("-x");
+        boolean matchesPattern = shouldMatchEntireLine
+            ? lineToCompare.equals(patternToCompare)
+            : lineToCompare.contains(patternToCompare);
+
+        boolean shouldInvertMatch = flags.contains("-v");
+        //noinspection SimplifiableConditionalExpression
+        return shouldInvertMatch ? !matchesPattern : matchesPattern;
```

Comment:
I know this can be simplified, but I think it's a lot easier to read like this. Happy to change it if people want though! 🙂

Reply:
I'm happy to keep this as it is - I like readability! :slightly_smiling_face: